File: app/src/main/java/com/asmat/rolando/bakingapp/widget/GroceriesListWidgetConfigureActivity.java
package com.asmat.rolando.bakingapp.widget;

import android.app.Activity;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.asmat.rolando.bakingapp.R;
import com.asmat.rolando.bakingapp.fragments.RecipesListFragment;
import com.asmat.rolando.bakingapp.models.Recipe;

/**
 * Created by rolandoasmat on 8/15/17.
 */

public class GroceriesListWidgetConfigureActivity extends AppCompatActivity implements RecipesListFragment.OnRecipeClickListener {
    private int mAppWidgetId;
    public static final String PREFS_NAME = "com.asmat.rolando.bakingapp.widget.GroceriesListWidget";
    public static final String PREF_PREFIX_KEY = "appwidget_";

    public void onCreate(Bundle icicle) {
        super.onCreate(icicle);
        this.setResult(Activity.RESULT_CANCELED);
        this.setContentView(R.layout.groceries_list_widget_configure);
        Bundle extras = getIntent().getExtras();
        if(extras != null) {
            this.mAppWidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        }

        if(this.mAppWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            finish();
        }
    }

    @Override
    public void onRecipeSelected( Recipe recipe) {
        Context context = getApplicationContext();
        saveTitlePref(context, mAppWidgetId, recipe.getName());

        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(this);
        GroceriesListWidget.updateAppWidget(context, appWidgetManager, mAppWidgetId);


        Intent resultValue = new Intent();
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId);
        setResult(Activity.RESULT_OK, resultValue);
        finish();
    }

    public static final void saveTitlePref( Context context, int appWidgetId,  String text) {
        SharedPreferences.Editor prefs = context.getSharedPreferences(PREFS_NAME, 0).edit();
        prefs.putString(PREF_PREFIX_KEY + appWidgetId, text);
        prefs.apply();
    }


    public static String loadTitlePref( Context context, int appWidgetId) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, 0);
        String titleValue = prefs.getString(PREF_PREFIX_KEY + appWidgetId, null);
        if(titleValue != null) {
            return titleValue;
        } else {
            return context.getString(R.string.appwidget_text);
        }
    }

    public static void deleteTitlePref( Context context, int appWidgetId) {
        SharedPreferences.Editor prefs = context.getSharedPreferences(PREFS_NAME, 0).edit();
        prefs.remove(PREF_PREFIX_KEY + appWidgetId);
        prefs.apply();
    }
}

File: app/src/main/java/com/asmat/rolando/bakingapp/widget/GroceriesListWidgetConfigureActivity.kt
package com.asmat.rolando.bakingapp.widget

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.asmat.rolando.bakingapp.R
import com.asmat.rolando.bakingapp.fragments.RecipesListFragment
import com.asmat.rolando.bakingapp.models.Recipe

/**
 * The configuration screen for the [GroceriesListWidget] AppWidget.
 */
class GroceriesListWidgetConfigureActivity : AppCompatActivity(), RecipesListFragment.OnRecipeClickListener {


    internal var mAppWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID

    public override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)

        // Set the result to CANCELED.  This will cause the widget host to cancel
        // out of the widget placement if the user presses the back button.
        setResult(Activity.RESULT_CANCELED)

        setContentView(R.layout.groceries_list_widget_configure)
//        mAppWidgetText = findViewById(R.id.appwidget_text) as EditText
//        findViewById(R.id.add_button).setOnClickListener(mOnClickListener)

        // Find the widget id from the intent.
        val intent = intent
        val extras = intent.extras
        if (extras != null) {
            mAppWidgetId = extras.getInt(
                    AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
        }

        // If this activity was started with an intent without an app widget ID, finish with an error.
        if (mAppWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            finish()
            return
        }
    }

    override fun onRecipeSelected(recipe: Recipe) {
        val context = this@GroceriesListWidgetConfigureActivity

        // When the button is clicked, store the string locally
        saveTitlePref(context, mAppWidgetId, recipe.name)

        // It is the responsibility of the configuration activity to update the app widget
        val appWidgetManager = AppWidgetManager.getInstance(context)
        GroceriesListWidget.updateAppWidget(context, appWidgetManager, mAppWidgetId)

        // Make sure we pass back the original appWidgetId
        val resultValue = Intent()
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId)
        setResult(Activity.RESULT_OK, resultValue)
        finish()
    }

    companion object {

        private val PREFS_NAME = "com.asmat.rolando.bakingapp.widget.GroceriesListWidget"
        private val PREF_PREFIX_KEY = "appwidget_"

        // Write the prefix to the SharedPreferences object for this widget
        internal fun saveTitlePref(context: Context, appWidgetId: Int, text: String) {
            val prefs = context.getSharedPreferences(PREFS_NAME, 0).edit()
            prefs.putString(PREF_PREFIX_KEY + appWidgetId, text)
            prefs.apply()
        }

        // Read the prefix from the SharedPreferences object for this widget.
        // If there is no preference saved, get the default from a resource
        internal fun loadTitlePref(context: Context, appWidgetId: Int): String {
            val prefs = context.getSharedPreferences(PREFS_NAME, 0)
            val titleValue = prefs.getString(PREF_PREFIX_KEY + appWidgetId, null)
            if (titleValue != null) {
                return titleValue
            } else {
                return context.getString(R.string.appwidget_text)
            }
        }

        internal fun deleteTitlePref(context: Context, appWidgetId: Int) {
            val prefs = context.getSharedPreferences(PREFS_NAME, 0).edit()
            prefs.remove(PREF_PREFIX_KEY + appWidgetId)
            prefs.apply()
        }
    }
}

